src/tsserver/nodeServer.ts

```diff
@@ -715,6 +715,42 @@ namespace ts.server {
                 this.constructed = true;
             }
 
+            public send(msg: protocol.Message) {
+                if (useNodeIpc) {

```

Comment:
Please consider refactoring. The repeated flag check is a good indicator that you actually want method overrides. Maybe we're looking at a third subclass of `Session` (`WorkerSession`, `IOSession` and maybe `NodeIpcSession`?)

`parseMessage` and `toStringMessage` are the same as in the `WorkerSession` implementation. 

`send` seems to be different for all three implementations, but does it need to be? That `msg.type === "event"` check and logging code applies to all implementations so it may as well live in the base class. Then you probably want to pull out the `writeMessage` call into a `protected` method to make it specific to the subclass, since that's the part that actually differs between the three.

(I don't know if we want to follow these OOP patterns all across the codebase, but since the session code is written that way, my advice is following that pattern.).

Reply:
Good suggestion. Made the fix